MigrationTutorial/Migrations/V2Utils.cs

```diff
@@ -0,0 +1,142 @@
+                    {
+                        Name = "Prototyping",
+                        Head = headDep2
+                    }
+                });
+
+                realm.Add(new Customer[]{
+                    new Customer()
+                    {
+                        Name = "Nuke",
+                        Location = "United States"
+                    },
+                    new Customer()
+                    {
+                        Name = "Gacci",
+                        Location = "Italy"
+                    }
+                });
+            });
+
+            var suppliers = realm.All<Supplier>();
+            var manufactDep = realm.All<Department>().Where(d => d.Name == "Manufacturing").First();
+            var consumables = realm.All<Consumable>();
+
+            realm.Write(() =>
+            {
+                for (var i = 0; i < employees.Count(); i++)
+                {
+                    employees.ElementAt(i).Department = manufactDep;
+                }
+
+                for (var i = 0; i < consumables.Count(); i++)
+                {
+                    // it is expected to have a supplier for everything
+                    var typeToSearch = consumables.ElementAt(i).Type;
+                    var matchingSupplier = suppliers.Filter("ANY _SuppliedTypes = $0", typeToSearch.ToString()).First();
+                    consumables.ElementAt(i).Supplier = matchingSupplier;
+                }
+            });
+        }
+
+        public static void DoMigrate(Migration migration)
+        {
+            var newEmployees = migration.NewRealm.All<Employee>();
+            var oldEmployees = migration.OldRealm.DynamicApi.All("Employee");
+
+            for (var i = 0; i < newEmployees.Count(); i++)
+            {
+                var newEmployee = newEmployees.ElementAt(i);
+                var oldEmployee = oldEmployees.ElementAt(i);
+                newEmployee.Gender = string.Equals(oldEmployee.Gender, "female", StringComparison.OrdinalIgnoreCase) ?
+                    Gender.Female : Gender.Male;
+            }
+
+            var newConsumables = migration.NewRealm.All<Consumable>();
+            var oldConsumables = migration.OldRealm.DynamicApi.All("Consumable");
+            var distinctConsumableId = new HashSet<string>();
+            var consumableToDelete = new List<Consumable>();
+            for (var i = 0; i < newConsumables.Count(); i++)
+            {
+                var currConsumable = newConsumables.ElementAt(i);
+
+                // remove duplicates since ProductId is the new PrimaryKey
+                if (distinctConsumableId.Contains(currConsumable.ProductId))
+                {
+                    consumableToDelete.Add(currConsumable);
+                }
+                else
+                {
+                    distinctConsumableId.Add(currConsumable.ProductId);
+                }
+            }
+
+            for (var i = 0; i < newConsumables.Count(); i++)
```

Comment:
Can't we merge the two for loops? We can set the last purchased price before.
Also, I suppose that deleting the elements of "consumableToDelete" needs to happen once, and not in a loop. 

Also, a question. What happens if you don't delete the values? Now we have a property that is a primary key ("ProductId"), that wasn't necessarily unique before. Would it give an error after the migration is done if we don't remove the elements with the same product id??

Reply:
> Can't we merge the two for loops? We can set the last purchased price before.

You're right. Fixed.

> Also, I suppose that deleting the elements of "consumableToDelete" needs to happen once, and not in a loop.

Some leftover since I did not use LINQ before. Fixed.

> Also, a question. What happens if you don't delete the values? Now we have a property that is a primary key ("ProductId"), that wasn't necessarily unique before. Would it give an error after the migration is done if we don't remove the elements with the same product id??

Good question. I did not try that, since it's logically wrong to have multiple objects with the same PK, I just removed those. Later I'm going to check if we have tests in our SDK for non unique PKs.